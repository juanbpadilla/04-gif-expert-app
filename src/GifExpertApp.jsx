import { useState } from "react"

export const GifExpertApp = () => {

  /**
   * ⚠️ IMPORTANTE:
   * Los hooks son funciones que nos permiten conectar el estado y el ciclo de vida de los componentes funcionales.
   * estos son pocisionales, es decir, el orden en el que se declaran es importante.
   * no se pueden llamar condicionalmente, ni dentro de un loop. Solo se pueden llamar en el nivel más alto de un componente.
   */
  const [ categories, setCategories ] = useState([ 'One Punch', 'Dragon Ball' ]);

  console.log(categories);

  return (
    <>
      {/* título */}
      <h1>GifExpertApp</h1>

      {/* Input */}

      {/* Listado de Gif */}
      <ol>
        { 
          categories.map( category => {
            return <li key={ category }>{ category }</li>
          }) 
        }
      </ol>
        {/* Gif Item */}

    </>
  )
}
