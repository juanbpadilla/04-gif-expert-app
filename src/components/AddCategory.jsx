import { useState } from "react";
import PropTypes from 'prop-types'

/**
 * setCategories es una función que se pasa como argumento al componente.
 * esta función se encarga de agregar un nuevo elemento al arreglo de categorías.
 * 
 * el componente AddCategory recibe un objeto con la propiedad setCategories.
 * esta propiedad es una función que se encarga de agregar un nuevo elemento al arreglo de categorías.
 * 
 * setCategories nos marca un error porque no está definido en el componente AddCategory.
 * para solucionar este error, debemos agregar setCategories como argumento de la función AddCategory.
 * con propTypes podemos definir las propiedades que recibe un componente.
 * 
 * @param {*} param0 
 * @returns 
 */
export const AddCategory = ({ setCategories }) => {

  const [ inputValue, setInputValue ] = useState('');

  /**
   * esta función se encarga de capturar el valor del input en tiempo real.
   * event.target.value nos permite obtener el valor del input.
   * 
   * también podemos desestructurar el evento y obtener el target directamente.
   * es decir, en lugar de (event) => onInputChange(event) podemos hacer ({ target }) => onInputChange(target)
   * 
   * @param {Event} event
   */
  const onInputChange = ({ target }) => {
    setInputValue( target.value );
  }

  /**
   * event.preventDefault() evita que el formulario se envíe automáticamente.
   * 
   * también podemos validar que el valor del input no esté vacío.
   * si el valor del input está vacío, retornamos un return para que la función no haga nada.
   * 
   * para agregar un nuevo elemento al arreglo, podemos usar el operador spread.
   * primero agregamos el nuevo valor y luego el resto de los elementos del arreglo.
   * prevState es el estado anterior del arreglo. (no necesariamente se debe llamar prevState)
   * setCategories(prevState => [ inputValue, ...prevState ]);
   * 
   * finalmente, limpiamos el input.
   * setInputValue('');
   * 
   * @param {*} event 
   */
  const onSubmit = ( event ) => {
    event.preventDefault();
    if ( inputValue.trim().length <= 1 ) return;

    setCategories(prevState => [ inputValue, ...prevState ]);
    setInputValue('');
  }

  /**
   * onSubmit es una función que se ejecuta cuando se envía el formulario.
   * 
   * onInputChange es una función que se ejecuta cuando se cambia el valor del input.
   * 
   * cuando en una función de flecha el primer argumento es el mismo
   * que se pasa a otra función, podemos simplificarlo de la siguiente forma:
   * De: { (event) => onInputChange(event) }
   * A:  { onInputChange }
   */
  return (
    <form onSubmit={ onSubmit }>
      <input
        type="text"
        placeholder="Buscar gifs"
        value={ inputValue }
        onChange={ onInputChange }
      />
    </form>
  )
}

AddCategory.propTypes = {
  setCategories: PropTypes.func.isRequired
}