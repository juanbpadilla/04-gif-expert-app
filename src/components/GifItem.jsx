import PropTypes from 'prop-types';

/**
 * GifItem recibe un gif con las propiedades title y url.
 * ..y se encarga de mostrar la imagen y el tÃ­tulo del gif.
 * 
 * @param {*} param0 
 * @returns 
 */
export const GifItem = ({ title, url }) => {
  return (
    <div className="card">
      <img src={ url } alt={ title } />
      <p>{ title }</p>
    </div>
  )
}

GifItem.propTypes = {
  title: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
};

