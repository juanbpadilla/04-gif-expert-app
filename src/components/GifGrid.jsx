import PropTypes from 'prop-types'

import { GifItem } from './GifItem';
import { useFetchGifs } from '../hooks/useFetchGifs';


/**
 * GifGrid recibe una categoría como propiedad y muestra los gifs de esa categoría.
 * 
 * En este punto estamos usando el custom hook "useFetchGifs" para obtener los gifs de la categoría,
 * este recibe una categoría como parámetro y se encarga de obtener los gifs de esa categoría, 
 * "useFetchGifs" retorna un objeto con las propiedades images y isLoading.
 * 
 * Se hace uso del componente "GifItem" para mostrar los gifs.
 * "GifItem" recibe como propiedad un gif con las propiedades id, title y url.  
 * Usamos el operador spread para pasar todas las propiedades del gif como propiedades independientes
 * y este componente se encarga de mostrar la imagen y el título del gif.
 * 
 * @param {string} category 
 * @returns 
 */
export const GifGrid = ({ category }) => {

  const { images, isLoading } = useFetchGifs( category );  

  return (
    <>
      <h3>{ category }</h3>

      <div className='card-grid'>
        {
          /**
           * images es un arreglo de gifs que se obtiene de la API de Giphy.
           * ..usamos el método map para recorrer el arreglo de gifs y retornar un componente GifItem por cada gif.
           * 
           * GifItem recibe como propiedad un gif con las propiedades id, title y url.
           * ..usamos el operador spread para pasar todas las propiedades del gif como propiedades independientes.
           * ..a esto se le conoce como "propiedades dinámicas" o "esparcimiento de propiedades". { ...image }
           * ..esto nos permite usar las propiedades del gif como propiedades independientes en el componente GifItem.
           * { ...image } es equivalente a { id={ image.id } title={ image.title } url={ image.url } }
           * 
           * @param {*} image
           * @returns
           */
          images.map(( image ) => (
            <GifItem 
              key={ image.id }
              { ...image }
            />
          ))
        }
      </div>
    </>
  )
}

GifGrid.propTypes = {
  category: PropTypes.string.isRequired
}