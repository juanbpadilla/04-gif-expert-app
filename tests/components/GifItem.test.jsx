import { render } from "@testing-library/react"
import { GifItem } from "../../src/components"

describe('pruebas en componente <GifItem />', () => {

  const title = 'Un tÃ­tulo';
  const url = 'https://localhost/algo.jpg';

  test('debe mostrar el componente correctamente', () => {
    
    const { container } = render(<GifItem title={ title } url={ url } />);
    expect( container ).toMatchSnapshot()

  })
})