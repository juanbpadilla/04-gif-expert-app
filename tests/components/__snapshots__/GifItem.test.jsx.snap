// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en componente <GifItem /> debe mostrar el componente correctamente 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="Un título"
      src="https://localhost/algo.jpg"
    />
    <p>
      Un título
    </p>
  </div>
</div>
`;
